### PR TITLE
nixos/adguardhome: run --check-config before merging

### LINKED ISSUE #246461: AdGuardHome does not start up after updating nixpkgs to e6ab469 (2023-07-29)
### Describe the bug

I recently updated my system and then AdGuardHome ceased to work.

```
• Updated input 'nixpkgs':
    'github:nixos/nixpkgs/12303c652b881435065a98729eb7278313041e49' (2023-07-22)
  → 'github:nixos/nixpkgs/e6ab46982debeab9831236869539a507f670a129' (2023-07-29)
```

The error message in the flake is:

```
adguardhome[22723]: 2023/07/31 19:17:16 [error] Couldn't get logging settings from the configuration: yaml: unmarshal errors:
adguardhome[22723]:   line 7: cannot unmarshal !!seq into filtering.BlockedServices
adguardhome[22723]:   line 28: cannot unmarshal !!seq into filtering.BlockedServices
adguardhome[22723]:   line 85: cannot unmarshal !!seq into filtering.BlockedServices
```

Originally I thought it was because /nix/store/*AdGuardHome.yaml is a JSON file, but it appears it is always that way. I'm not entirely sure what is missing to produce that error. I had a `services.adguardhome.settings` set up for rewriting, but commenting that out continued to produce the error so it doesn't seem to be that.

```
  services.adguardhome = {
    enable = true;
    openFirewall = true;
    allowDHCP = true;
    settings.bind_host = "0.0.0.0";
    settings.bind_port = port;
  };
```

### Steps To Reproduce

1. Set up AdGuardHome using the above stanza.
2. Start up the service.

### Expected behavior

It would start up and protecting my location from the evils of surveillance capitalism.

### Screenshots

N/A

### Additional context

N/A

### Notify maintainers

- @r-ryantm 
- @numkem
- @iagoq
- @rhoriguchi

### Metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.37, NixOS, 23.11 (Tapir), 23.11pre-git`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.15.1`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

### PR DESCRIPTION
**Pull request by CRTified**
To prevent merging configurations on different `schema_version`, we will now check the existing configuration, which has the side-effect of updating the `schema_version`.

Fixes #246461 as described in https://github.com/NixOS/nixpkgs/issues/246461#issuecomment-1667663556

Should work/merge independent of #247828, which changes the generation of the default config.

## Description of changes

Add an additional call to `adguardhome --check-config` to the `preStart`.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [X] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by CRTified** [comment]
Pinging @dmoonfire as creator of the original issue, as well.

**Review by rhoriguchi — APPROVED** [review]

File: nixos/modules/services/networking/adguardhome.nix [hunk 1]
```diff
@@ -150,6 +150,11 @@ in
       preStart = optionalString (cfg.settings != null) ''
         if    [ -e "$STATE_DIRECTORY/AdGuardHome.yaml" ] \
            && [ "${toString cfg.mutableSettings}" = "1" ]; then
+          # First run a schema_version update on the existing configuration
+          # This ensures that both the new config and the existing one have the same schema_version
+          # Note: --check-config has the side effect of modifying the file at rest!
+          ${pkgs.adguardhome}/bin/adguardhome -c "$STATE_DIRECTORY/AdGuardHome.yaml" --check-config
```
Comment (bjornfor) [inline nixos/modules/services/networking/adguardhome.nix:156]: `${pkgs.adguardhome}/bin/adguardhome` -> `${lib.getExe adguardhome}`?

Comment (CRTified) [reply nixos/modules/services/networking/adguardhome.nix]: Also fixed the other occurrences. Tests pass for me on `x86_64-linux`.

**Comment by rhoriguchi** [comment]
@ofborg eval

**Review by rhoriguchi — APPROVED** [review]

**Comment by CRTified** [comment]
@wegank Hope it's fine to ping you, as you have merged the related #247828. This PR closes a related, repeatedly reported issue with AdGuard Home.

**Review by wegank — APPROVED** [review]

File: nixos/modules/services/networking/adguardhome.nix [hunk 2]
```diff
@@ -167,8 +167,13 @@ in {
       preStart = optionalString (settings != null) ''
         if    [ -e "$STATE_DIRECTORY/AdGuardHome.yaml" ] \
            && [ "${toString cfg.mutableSettings}" = "1" ]; then
+          # First run a schema_version update on the existing configuration
+          # This ensures that both the new config and the existing one have the same schema_version
+          # Note: --check-config has the side effect of modifying the file at rest!
+          ${lib.getExe pkgs.adguardhome} -c "$STATE_DIRECTORY/AdGuardHome.yaml" --check-config
```
Comment (wegank) [inline nixos/modules/services/networking/adguardhome.nix:173]: Should `pkgs.adguardhome` be `cfg.package` instead?

Comment (CRTified) [reply nixos/modules/services/networking/adguardhome.nix:173]: True. Give me a second.

**Review by rhoriguchi — APPROVED** [review]

**Pull request merged**